packages/mui-base/src/RadioGroup/Indicator/RadioGroupIndicator.tsx

```diff
@@ -0,0 +1,78 @@
+'use client';
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { useComponentRenderer } from '../../utils/useComponentRenderer';
+import type { CustomStyleHookMapping } from '../../utils/getStyleHookProps';
+import { useRadioGroupItemContext } from '../Item/RadioGroupItemContext';
+import type {
+  RadioGroupIndicatorOwnerState,
+  RadioGroupIndicatorProps,
+} from './RadioGroupIndicator.types';
+
+const customStyleHookMapping: CustomStyleHookMapping<RadioGroupIndicatorOwnerState> = {
+  checked(value) {
+    return {
+      'data-state': value ? 'checked' : 'unchecked',
+    };
+  },
+};
+
+const RadioGroupIndicator = React.forwardRef(function RadioGroupIndicator(
+  props: RadioGroupIndicatorProps,
+  forwardedRef: React.ForwardedRef<HTMLSpanElement>,
+) {
+  const { render, className, keepMounted = false, ...otherProps } = props;
+
+  const { disabled, checked, required, readOnly } = useRadioGroupItemContext();
+
+  const ownerState: RadioGroupIndicatorOwnerState = React.useMemo(
+    () => ({
+      disabled,
+      checked,
+      required,
+      readOnly,
+    }),
+    [disabled, checked, required, readOnly],
+  );
+
+  const { renderElement } = useComponentRenderer({
+    render: render ?? 'span',
+    ref: forwardedRef,
+    className,
+    ownerState,
+    extraProps: otherProps,
+    customStyleHookMapping,
+  });
+
+  if (!keepMounted && !ownerState.checked) {
+    return null;
```

Comment:
I feel like an indicator that renders only when a radio is checked covers a pretty specific use case. Radio buttons usually have different indicators for checked and unchecked states. Since this is purely presentational, do we even need a component for it? If we decide to keep it, I'd vote for having it always rendered and placing a data-state attribute on it. I'd also consider accepting a child function to be able to render a different icon based on state.

Reply:
This is fair point. It has the same API as `Checkbox.Root` which unmounts when not checked. But for Radio Groups, often you want to keep an indicator even when unchecked (in fact, that's how the demos work). I think keeping it but using an attribute to style seems ok, but I wonder if we should change Checkbox just for consistency too